File: plugin/src/knowledge/providers/github.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const GITHUB_API = "https://api.github.com";

interface GitHubRepoResult {
  items?: Array<{
    id: number;
    name: string;
    full_name: string;
    description: string | null;
    html_url: string;
    stargazers_count: number;
    forks_count: number;
    language: string | null;
    topics?: string[];
    owner: {
      login: string;
      avatar_url: string;
    };
    updated_at: string;
    license?: {
      spdx_id: string;
    };
  }>;
  total_count?: number;
}

interface GitHubUserResult {
  items?: Array<{
    id: number;
    login: string;
    avatar_url: string;
    html_url: string;
    type: string;
  }>;
}

interface GitHubUser {
  login: string;
  name: string | null;
  bio: string | null;
  html_url: string;
  public_repos: number;
  followers: number;
  following: number;
  company: string | null;
  location: string | null;
  blog: string | null;
  twitter_username: string | null;
}

interface GitHubRepo {
  name: string;
  full_name: string;
  description: string | null;
  html_url: string;
  stargazers_count: number;
  forks_count: number;
  open_issues_count: number;
  language: string | null;
  topics?: string[];
  license?: { spdx_id: string };
  created_at: string;
  updated_at: string;
  owner: { login: string };
}

interface GitHubConfig {
  token?: string;
}

export class GitHubProvider implements KnowledgeProvider {
  readonly name = "github";
  private token?: string;

  configure(config: GitHubConfig): void {
    this.token = config.token;
  }

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      Accept: "application/vnd.github.v3+json",
      "User-Agent": "VaultTree/0.1.0",
    };
    if (this.token) {
      headers["Authorization"] = `Bearer ${this.token}`;
    }
    return headers;
  }

  async isAvailable(): Promise<boolean> {
    try {
      const response = await requestUrl({
        url: `${GITHUB_API}/rate_limit`,
        method: "GET",
        headers: this.getHeaders(),
        throw: false,
      });
      return response.status === 200;
    } catch {
      return false;
    }
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    try {
      // Check if query is a GitHub URL
      const urlMatch = this.parseGitHubUrl(query);
      if (urlMatch) {
        const entry = await this.lookupByUrl(urlMatch);
        return {
          success: true,
          provider: this.name,
          entries: entry ? [entry] : [],
        };
      }

      // Check if query looks like owner/repo
      if (query.includes("/") && !query.includes(" ")) {
        const entry = await this.lookupRepo(query);
        if (entry) {
          return {
            success: true,
            provider: this.name,
            entries: [entry],
          };
        }
      }

      // Otherwise, search repos
      const entries = await this.searchRepos(query, maxResults);

      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `GitHub lookup failed: ${message}`,
      };
    }
  }

  private parseGitHubUrl(query: string): { type: "repo" | "user"; path: string } | null {
    const repoMatch = /github\.com\/([^\/]+\/[^\/]+)\/?/.exec(query);
    if (repoMatch) {
      return { type: "repo", path: repoMatch[1] };
    }

    const userMatch = /github\.com\/([^\/]+)\/?$/.exec(query);
    if (userMatch) {
      return { type: "user", path: userMatch[1] };
    }

    return null;
  }

  private async lookupByUrl(match: { type: "repo" | "user"; path: string }): Promise<KnowledgeEntry | null> {
    if (match.type === "repo") {
      return this.lookupRepo(match.path);
    } else {
      return this.lookupUser(match.path);
    }
  }

  private async lookupRepo(fullName: string): Promise<KnowledgeEntry | null> {
    const response = await requestUrl({
      url: `${GITHUB_API}/repos/${fullName}`,
      method: "GET",
      headers: this.getHeaders(),
      throw: false,
    });

    if (response.status === 404) {
      return null;
    }

    if (response.status !== 200) {
      throw new Error(`Repo lookup failed: ${response.status}`);
    }

    const repo = response.json as GitHubRepo;
    return this.repoToEntry(repo);
  }

  private async lookupUser(username: string): Promise<KnowledgeEntry | null> {
    const response = await requestUrl({
      url: `${GITHUB_API}/users/${username}`,
      method: "GET",
      headers: this.getHeaders(),
      throw: false,
    });

    if (response.status === 404) {
      return null;
    }

    if (response.status !== 200) {
      throw new Error(`User lookup failed: ${response.status}`);
    }

    const user = response.json as GitHubUser;

    const lines: string[] = [];
    if (user.bio) lines.push(user.bio);
    lines.push(`Repos: ${user.public_repos} | Followers: ${user.followers}`);
    if (user.company) lines.push(`Company: ${user.company}`);
    if (user.location) lines.push(`Location: ${user.location}`);

    return {
      title: user.name || user.login,
      summary: lines.join("\n"),
      url: user.html_url,
      source: "github",
      metadata: {
        type: "user",
        login: user.login,
        name: user.name,
        bio: user.bio,
        publicRepos: user.public_repos,
        followers: user.followers,
        following: user.following,
        company: user.company,
        location: user.location,
        blog: user.blog,
        twitter: user.twitter_username,
      },
    };
  }

  private async searchRepos(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const params = new URLSearchParams({
      q: query,
      sort: "stars",
      order: "desc",
      per_page: String(limit),
    });

    const response = await requestUrl({
      url: `${GITHUB_API}/search/repositories?${params}`,
      method: "GET",
      headers: this.getHeaders(),
      throw: false,
    });

    if (response.status !== 200) {
      throw new Error(`Search failed: ${response.status}`);
    }

    const data = response.json as GitHubRepoResult;

    if (!data.items) {
      return [];
    }

    return data.items.map((repo) => this.repoToEntry(repo));
  }

  private repoToEntry(repo: GitHubRepo | GitHubRepoResult["items"][0]): KnowledgeEntry {
    const stars = repo.stargazers_count >= 1000
      ? `${(repo.stargazers_count / 1000).toFixed(1)}k`
      : String(repo.stargazers_count);

    const forks = repo.forks_count >= 1000
      ? `${(repo.forks_count / 1000).toFixed(1)}k`
      : String(repo.forks_count);

    const lines: string[] = [];
    if (repo.description) lines.push(repo.description);
    lines.push(`Stars: ${stars} | Forks: ${forks}`);
    if (repo.language) lines.push(`Language: ${repo.language}`);
    if (repo.topics && repo.topics.length > 0) {
      lines.push(`Topics: ${repo.topics.slice(0, 5).join(", ")}`);
    }

    return {
      title: repo.full_name,
      summary: lines.join("\n"),
      url: repo.html_url,
      source: "github",
      metadata: {
        type: "repo",
        name: repo.name,
        fullName: repo.full_name,
        owner: repo.owner.login,
        description: repo.description,
        stars: repo.stargazers_count,
        forks: repo.forks_count,
        language: repo.language,
        topics: repo.topics,
        license: repo.license?.spdx_id,
        updatedAt: repo.updated_at,
      },
    };
  }
}

File: crates/mcp/src/knowledge/github.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::collections::HashMap;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const GITHUB_API: &str = "https://api.github.com";

pub struct GitHubProvider {
    client: Client,
    token: Option<String>,
}

#[derive(Debug, Deserialize)]
struct RepoSearchResponse {
    items: Option<Vec<RepoItem>>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct RepoItem {
    full_name: String,
    name: String,
    description: Option<String>,
    html_url: String,
    stargazers_count: u64,
    forks_count: u64,
    language: Option<String>,
    topics: Option<Vec<String>>,
    license: Option<License>,
    updated_at: Option<String>,
    owner: Owner,
}

#[derive(Debug, Deserialize)]
struct Owner {
    login: String,
}

#[derive(Debug, Deserialize)]
struct License {
    spdx_id: Option<String>,
}

impl GitHubProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
            token: None,
        }
    }

    pub fn with_token(token: impl Into<String>) -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
            token: Some(token.into()),
        }
    }

    fn add_auth(&self, request: reqwest::blocking::RequestBuilder) -> reqwest::blocking::RequestBuilder {
        match &self.token {
            Some(token) => request.header("Authorization", format!("Bearer {}", token)),
            None => request,
        }
    }

    fn format_count(n: u64) -> String {
        if n >= 1000 {
            format!("{:.1}k", n as f64 / 1000.0)
        } else {
            n.to_string()
        }
    }

    fn repo_to_entry(&self, repo: &RepoItem) -> KnowledgeEntry {
        let stars = Self::format_count(repo.stargazers_count);
        let forks = Self::format_count(repo.forks_count);

        let mut lines = Vec::new();
        if let Some(desc) = &repo.description {
            lines.push(desc.clone());
        }
        lines.push(format!("Stars: {} | Forks: {}", stars, forks));
        if let Some(lang) = &repo.language {
            lines.push(format!("Language: {}", lang));
        }
        if let Some(topics) = &repo.topics {
            if !topics.is_empty() {
                lines.push(format!("Topics: {}", topics.iter().take(5).cloned().collect::<Vec<_>>().join(", ")));
            }
        }

        let mut metadata = HashMap::new();
        metadata.insert("type".to_string(), serde_json::json!("repo"));
        metadata.insert("name".to_string(), serde_json::json!(repo.name));
        metadata.insert("fullName".to_string(), serde_json::json!(repo.full_name));
        metadata.insert("owner".to_string(), serde_json::json!(repo.owner.login));
        if let Some(desc) = &repo.description {
            metadata.insert("description".to_string(), serde_json::json!(desc));
        }
        metadata.insert("stars".to_string(), serde_json::json!(repo.stargazers_count));
        metadata.insert("forks".to_string(), serde_json::json!(repo.forks_count));
        if let Some(lang) = &repo.language {
            metadata.insert("language".to_string(), serde_json::json!(lang));
        }
        if let Some(topics) = &repo.topics {
            metadata.insert("topics".to_string(), serde_json::json!(topics));
        }
        if let Some(lic) = &repo.license {
            if let Some(id) = &lic.spdx_id {
                metadata.insert("license".to_string(), serde_json::json!(id));
            }
        }

        KnowledgeEntry {
            title: repo.full_name.clone(),
            summary: lines.join("\n"),
            url: Some(repo.html_url.clone()),
            source: "github".to_string(),
            metadata: Some(metadata),
        }
    }

    fn search_repos(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/search/repositories?q={}&sort=stars&order=desc&per_page={}",
            GITHUB_API,
            urlencoding::encode(query),
            limit
        );

        let request = self.client
            .get(&url)
            .header("Accept", "application/vnd.github.v3+json");

        let response = self.add_auth(request)
            .send()
            .map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Err(format!("search failed: {}", response.status()));
        }

        let data: RepoSearchResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .items
            .unwrap_or_default()
            .iter()
            .map(|r| self.repo_to_entry(r))
            .collect())
    }

    fn lookup_repo(&self, full_name: &str) -> Result<Option<KnowledgeEntry>, String> {
        let url = format!("{}/repos/{}", GITHUB_API, full_name);

        let request = self.client
            .get(&url)
            .header("Accept", "application/vnd.github.v3+json");

        let response = self.add_auth(request)
            .send()
            .map_err(|e| e.to_string())?;

        if response.status().as_u16() == 404 {
            return Ok(None);
        }

        if !response.status().is_success() {
            return Err(format!("repo lookup failed: {}", response.status()));
        }

        let repo: RepoItem = response.json().map_err(|e| e.to_string())?;
        Ok(Some(self.repo_to_entry(&repo)))
    }
}

impl Default for GitHubProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for GitHubProvider {
    fn name(&self) -> &'static str {
        "github"
    }

    fn is_available(&self) -> bool {
        let request = self.client
            .get(format!("{}/rate_limit", GITHUB_API))
            .header("Accept", "application/vnd.github.v3+json");

        self.add_auth(request)
            .send()
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }

    fn lookup(&self, query: &str, options: &LookupOptions) -> LookupResult {
        let limit = options.max_results.unwrap_or(5);

        // Check if query looks like owner/repo
        if query.contains('/') && !query.contains(' ') {
            match self.lookup_repo(query) {
                Ok(Some(entry)) => return LookupResult::success(self.name(), vec![entry]),
                Ok(None) => {}
                Err(e) => return LookupResult::error(self.name(), e),
            }
        }

        // Search repos
        match self.search_repos(query, limit) {
            Ok(entries) => LookupResult::success(self.name(), entries),
            Err(e) => LookupResult::error(self.name(), e),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore] // Requires network
    fn github_lookup() {
        let provider = GitHubProvider::new();
        let result = provider.lookup("rust-lang/rust", &LookupOptions::default());
        assert!(result.success);
        assert!(!result.entries.is_empty());
    }
}
